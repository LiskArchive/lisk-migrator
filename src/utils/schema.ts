/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export const defaultAccountSchema = {
	token: {
		type: 'object',
		fieldNumber: 2,
		properties: {
			balance: {
				fieldNumber: 1,
				dataType: 'uint64',
			},
		},
		default: {
			balance: BigInt(0),
		},
	},
	sequence: {
		type: 'object',
		fieldNumber: 3,
		properties: {
			nonce: {
				fieldNumber: 1,
				dataType: 'uint64',
			},
		},
		default: {
			nonce: BigInt(0),
		},
	},
	keys: {
		type: 'object',
		fieldNumber: 4,
		properties: {
			numberOfSignatures: { dataType: 'uint32', fieldNumber: 1 },
			mandatoryKeys: {
				type: 'array',
				items: { dataType: 'bytes' },
				fieldNumber: 2,
			},
			optionalKeys: {
				type: 'array',
				items: { dataType: 'bytes' },
				fieldNumber: 3,
			},
		},
		default: {
			numberOfSignatures: 0,
			mandatoryKeys: [],
			optionalKeys: [],
		},
	},
	dpos: {
		type: 'object',
		fieldNumber: 5,
		properties: {
			delegate: {
				type: 'object',
				fieldNumber: 1,
				properties: {
					username: { dataType: 'string', fieldNumber: 1 },
					pomHeights: {
						type: 'array',
						items: { dataType: 'uint32' },
						fieldNumber: 2,
					},
					consecutiveMissedBlocks: { dataType: 'uint32', fieldNumber: 3 },
					lastForgedHeight: { dataType: 'uint32', fieldNumber: 4 },
					isBanned: { dataType: 'boolean', fieldNumber: 5 },
					totalVotesReceived: { dataType: 'uint64', fieldNumber: 6 },
				},
				required: [
					'username',
					'pomHeights',
					'consecutiveMissedBlocks',
					'lastForgedHeight',
					'isBanned',
					'totalVotesReceived',
				],
			},
			sentVotes: {
				type: 'array',
				fieldNumber: 2,
				items: {
					type: 'object',
					properties: {
						delegateAddress: {
							dataType: 'bytes',
							fieldNumber: 1,
						},
						amount: {
							dataType: 'uint64',
							fieldNumber: 2,
						},
					},
					required: ['delegateAddress', 'amount'],
				},
			},
			unlocking: {
				type: 'array',
				fieldNumber: 3,
				items: {
					type: 'object',
					properties: {
						delegateAddress: {
							dataType: 'bytes',
							fieldNumber: 1,
						},
						amount: {
							dataType: 'uint64',
							fieldNumber: 2,
						},
						unvoteHeight: {
							dataType: 'uint32',
							fieldNumber: 3,
						},
					},
					required: ['delegateAddress', 'amount', 'unvoteHeight'],
				},
			},
		},
		default: {
			delegate: {
				username: '',
				pomHeights: [],
				consecutiveMissedBlocks: 0,
				lastForgedHeight: 0,
				isBanned: false,
				totalVotesReceived: BigInt(0),
			},
			sentVotes: [],
			unlocking: [],
		},
	},
};
