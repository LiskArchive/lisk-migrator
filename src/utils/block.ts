/*
 * Copyright Â© 2023 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { codec, Schema } from '@liskhq/lisk-codec';
import { Database } from '@liskhq/lisk-db';
import { BlockHeader } from '@liskhq/lisk-chain';

import { DB_KEY_BLOCKS_ID } from '../constants';
import { blockHeaderSchema } from '../schemas';
import { keyString, incrementOne } from './transaction';

export const getDataFromDBStream = async (stream: NodeJS.ReadableStream, schema: Schema) => {
	const data = await new Promise<Record<string, unknown>[]>((resolve, reject) => {
		const result: Record<string, unknown>[] = [];
		stream
			.on('data', async ({ value }) => {
				const decodedResult: Record<string, unknown> = await codec.decode(schema, value);
				result.push(decodedResult);
			})
			.on('error', error => {
				reject(error);
			})
			.on('end', () => {
				resolve(result);
			});
	});
	return data;
};

export const getBlockPublicKeySet = async (
	db: Database,
	pageSize: number,
): Promise<Set<string>> => {
	const result = new Set<string>();
	let startingKey = Buffer.from(`${DB_KEY_BLOCKS_ID}:${keyString(Buffer.alloc(32, 0))}`);
	// eslint-disable-next-line no-constant-condition
	while (true) {
		let exist = false;
		const blocksStream = db.createReadStream({
			gte: startingKey,
			lte: Buffer.from(`${DB_KEY_BLOCKS_ID}:${keyString(Buffer.alloc(32, 255))}`),
			limit: pageSize,
		});
		let lastKey = startingKey;
		// eslint-disable-next-line no-loop-func
		await new Promise<void>((resolve, reject) => {
			blocksStream
				.on('data', async ({ key, value }) => {
					exist = true;
					const header = await codec.decode<BlockHeader>(blockHeaderSchema, value);
					result.add(header.generatorPublicKey.toString('hex'));
					lastKey = key;
				})
				.on('error', error => {
					reject(error);
				})
				.on('end', () => {
					resolve();
				});
		});
		if (!exist) {
			break;
		}
		startingKey = incrementOne(lastKey as Buffer);
	}
	return result;
};
