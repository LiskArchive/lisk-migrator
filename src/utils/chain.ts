/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import cli from 'cli-ux';
import { getAPIClient } from '../client';
import { NETWORK_CONSTANT } from '../constants';

let tokenIDLsk: string;
let heightPreviousSnapshotBlock: number;

interface ObserveParams {
	readonly label: string;
	readonly height: number;
	readonly liskCorePath: string;
	readonly delay: number;
	readonly isFinal: boolean;
}

export const getTokenIDLsk = (): string => tokenIDLsk;

export const setTokenIDLskByNetID = async (networkIdentifier: string): Promise<void> => {
	tokenIDLsk = NETWORK_CONSTANT[networkIdentifier].tokenID as string;
};

export const getHeightPrevSnapshotBlock = (): number => heightPreviousSnapshotBlock;

export const setHeightPrevSnapshotBlockByNetID = async (
	networkIdentifier: string,
): Promise<void> => {
	heightPreviousSnapshotBlock = NETWORK_CONSTANT[networkIdentifier]
		.snapshotHeightPrevious as number;
};

export const getNodeInfo = async (
	liskCorePath: string,
): Promise<{ height: number; finalizedHeight: number }> => {
	const client = await getAPIClient(liskCorePath);
	const { height, finalizedHeight } = await client.node.getNodeInfo();
	return { height, finalizedHeight };
};

const secondsToHumanString = (seconds: number): string => {
	const years = Math.floor(seconds / 31536000);
	const days = Math.floor((seconds % 31536000) / 86400);
	const hours = Math.floor(((seconds % 31536000) % 86400) / 3600);
	const minutes = Math.floor((((seconds % 31536000) % 86400) % 3600) / 60);
	const numSeconds = (((seconds % 31536000) % 86400) % 3600) % 60;

	const result = [];

	if (years > 0) {
		result.push(`${years}y`);
	}

	if (days > 0) {
		result.push(`${days}d`);
	}

	if (hours > 0) {
		result.push(`${hours}h`);
	}

	if (minutes > 0) {
		result.push(`${minutes}m`);
	}

	if (numSeconds > 0) {
		result.push(`${numSeconds}s`);
	}

	if (result.length === 0) {
		return '0';
	}

	return result.join(' ');
};

const getRemainingTime = (currentHeight: number, observedHeight: number): string =>
	secondsToHumanString((observedHeight - currentHeight) * 10);

export const observeChainHeight = async (options: ObserveParams): Promise<number> => {
	const observedHeight = options.height;
	const startHeight = options.isFinal
		? (await getNodeInfo(options.liskCorePath)).finalizedHeight
		: (await getNodeInfo(options.liskCorePath)).height;

	if (startHeight >= observedHeight) {
		return startHeight;
	}

	const progress = cli.progress({
		format: `${options.label}: [{bar}] {percentage}% | Remaining: {remaining}/{total} | Height: {height}/${observedHeight} | ETA: {timeLeft}`,
		fps: 2,
		synchronousUpdate: false,
		etaAsynchronousUpdate: false,
		barsize: 30,
	});

	progress.start(observedHeight - startHeight, 0, {
		timeLeft: getRemainingTime(startHeight, observedHeight),
		remaining: observedHeight - startHeight,
		height: startHeight,
	});

	await new Promise((resolve, reject) => {
		let intervalId: NodeJS.Timer;

		// eslint-disable-next-line consistent-return
		const checkHeight = async () => {
			let height!: number;
			try {
				height = options.isFinal
					? (await getNodeInfo(options.liskCorePath)).finalizedHeight
					: (await getNodeInfo(options.liskCorePath)).height;
			} catch (error) {
				return reject(error);
			}

			progress.update(height - startHeight, {
				timeLeft: getRemainingTime(height, observedHeight),
				remaining: observedHeight - height,
				height,
			});

			if (height >= observedHeight) {
				clearInterval(intervalId);
				return resolve(height);
			}
		};

		intervalId = setInterval(checkHeight, options.delay);
	});

	progress.stop();

	return observedHeight;
};
