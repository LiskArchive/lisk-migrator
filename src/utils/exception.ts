/*
 * Copyright Â© 2023 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { ERROR_CODE } from '../constants';

export class MigratorException extends Error {
	public code: number = ERROR_CODE.DEFAULT;

	public constructor(message: string, code: number) {
		super(message);
		this.code = code;
	}
}
