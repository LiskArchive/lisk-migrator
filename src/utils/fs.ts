/*
 * Copyright Â© 2023 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import * as tar from 'tar';
import fs from 'fs';
import path from 'path';

export const extractTarBall = async (
	srcFilePath: string,
	destDirPath: string,
): Promise<boolean | Error> =>
	new Promise((resolve, reject) => {
		if (fs.existsSync(destDirPath)) fs.rmdirSync(destDirPath, { recursive: true });
		fs.mkdirSync(destDirPath, { recursive: true });

		const fileStream = fs.createReadStream(srcFilePath);
		fileStream.pipe(tar.extract({ cwd: destDirPath }));
		fileStream.on('error', err => reject(new Error(err)));
		// Adding delay of 100ms since the promise resolves earlier than expected
		fileStream.on('end', () => setTimeout(resolve.bind(null, true), 100));
	});

export const exists = async (inputPath: string): Promise<boolean | Error> => {
	try {
		await fs.promises.access(inputPath);
		return true;
	} catch (_) {
		return false;
	}
};

export const rmdir = async (directoryPath: string, options = {}): Promise<boolean | Error> =>
	new Promise((resolve, reject) => {
		fs.rmdir(directoryPath, options, err => {
			if (err) return reject(err);
			return resolve(true);
		});
	});

/* eslint-disable @typescript-eslint/no-unused-expressions */
export const copyDir = async (src: string, dest: string) => {
	await fs.promises.mkdir(dest, { recursive: true });
	const files = await fs.promises.readdir(src, { withFileTypes: true });

	for (const fileInfo of files) {
		const srcPath = path.join(src, fileInfo.name);
		const destPath = path.join(dest, fileInfo.name);

		fileInfo.isDirectory()
			? await copyDir(srcPath, destPath)
			: await fs.promises.copyFile(srcPath, destPath);
	}
};

export const write = async (filePath: string, content: string): Promise<boolean | Error> =>
	new Promise((resolve, reject) => {
		fs.writeFile(filePath, content, err => {
			if (err) {
				return reject(err);
			}
			return resolve(true);
		});
	});

export const read = async (filePath: string): Promise<string | Error> =>
	new Promise((resolve, reject) => {
		fs.promises
			.readFile(filePath, 'utf8')
			.then(data => {
				resolve(data);
			})
			.catch(error => {
				reject(error);
			});
	});

export const copyFile = async (src: string, dest: string): Promise<boolean | Error> =>
	new Promise((resolve, reject) => {
		fs.copyFile(src, dest, err => {
			if (err) {
				return reject(err);
			}
			return resolve(true);
		});
	});

export const createTarball = async (filePath: string, outputDir: string) =>
	new Promise((resolve, reject) => {
		const fileName = path.basename(filePath);

		tar
			.create(
				{
					gzip: true,
					file: `${outputDir}/${fileName}.tar.gz`,
					cwd: outputDir,
				},
				[fileName],
			)
			.then(() => resolve(true))
			.catch(err => reject(err));
	});

export const getFiles = async (directoryPath: string, options = {}): Promise<string[] | Error> =>
	new Promise((resolve, reject) => {
		fs.readdir(directoryPath, options, (err, files) => {
			if (err) {
				return reject(err);
			}
			return resolve(files as string[]);
		});
	});
