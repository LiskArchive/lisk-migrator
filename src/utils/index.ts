export * from './genesis_block';
export * from './config';
export * from './chain';
