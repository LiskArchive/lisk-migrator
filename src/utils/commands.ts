/*
 * Copyright Â© 2023 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable @typescript-eslint/restrict-template-expressions */
import { resolve } from 'path';
import Command from '@oclif/command';

import { read, write, exists } from './fs';
import { FILE_NAME } from '../constants';
import { NetworkConfigLocal } from '../types';

export const getCommandsToExecPostMigration = async (
	networkConstant: NetworkConfigLocal,
	snapshotHeight: number,
	outputDir: string,
) => {
	const commandsToExecute = [];

	const forgingStatusJsonFilepath = resolve(outputDir, FILE_NAME.FORGING_STATUS);
	if (await exists(forgingStatusJsonFilepath)) {
		const forgingStatusString = (await read(forgingStatusJsonFilepath)) as string;
		const forgingStatusJson = JSON.parse(forgingStatusString);

		const chainID = parseInt(networkConstant.tokenID.substring(0, 2), 16);

		for (const forgingStatus of forgingStatusJson) {
			commandsToExecute.push(
				'\n',
				`## Generate/Register the BLS keys for validator ${forgingStatus.lskAddress} - Please modify the command if necessary`,
				'\n',
			);

			const keysFilepath = resolve(outputDir, FILE_NAME.KEYS);
			commandsToExecute.push(
				`lisk-core keys:create --chainid ${chainID} --output ${keysFilepath} --add-legacy`,
				`lisk-core keys:import --file-path ${keysFilepath}`,
				`lisk-core endpoint:invoke random_setHashOnion '{ "address":"${forgingStatus.lskAddress}"}'`,
				`lisk-core endpoint:invoke generator_setStatus '{ "address":"${
					forgingStatus.lskAddress
				}", "height": ${forgingStatus.height ?? snapshotHeight}, "maxHeightGenerated":  ${
					forgingStatus.maxHeightPreviouslyForged ?? snapshotHeight
				}, "maxHeightPrevoted":  ${forgingStatus.maxHeightPrevoted ?? snapshotHeight} }' --pretty`,
				`lisk-core generator:enable ${forgingStatus.lskAddress} --use-status-value`,
				'lisk-core transaction:create legacy registerKeys 400000 --key-derivation-path=legacy --send',
			);

			commandsToExecute.push('\n', '-----------------------------------------------------', '\n');
		}
	}

	return commandsToExecute;
};

export const writeCommandsToExec = async (
	_this: Command,
	networkConstant: NetworkConfigLocal,
	snapshotHeight: number,
	outputDir: string,
	preCompletionCommands?: string[],
) => {
	const commandsToExecPreCompletion = preCompletionCommands ?? [];
	const commandsToExecPostMigration = await getCommandsToExecPostMigration(
		networkConstant,
		snapshotHeight,
		outputDir,
	);

	const allCommandsToExec = [
		'## Please execute the following commands to finish the migration successfully:',
		'------------------------------------------------------------------------------',
		'\n',
		...commandsToExecPreCompletion,
		...commandsToExecPostMigration,
	].join('\n');

	// Create the document only when there are commands to be executed by the user
	if (commandsToExecPostMigration.length || commandsToExecPreCompletion.length) {
		const commandsToExecuteFilepath = resolve(outputDir, FILE_NAME.COMMANDS_TO_EXEC);

		_this.log(
			commandsToExecPreCompletion.length
				? `Creating file with the list of commands to execute post migration: ${commandsToExecuteFilepath}`
				: `Creating file with the list of commands to execute: ${commandsToExecuteFilepath}`,
		);

		await write(commandsToExecuteFilepath, allCommandsToExec);

		_this.log(
			commandsToExecPreCompletion.length
				? `Successfully created file with the list of commands to execute post migration: ${commandsToExecuteFilepath}`
				: `Successfully created file with the list of commands to execute: ${commandsToExecuteFilepath}`,
		);
	}
};
