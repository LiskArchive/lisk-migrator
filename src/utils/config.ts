/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable no-param-reassign */
import debugInit from 'debug';
import cli from 'cli-ux';
import { existsSync, readdirSync, mkdirSync, writeFileSync, unlinkSync } from 'fs';
import { execSync } from 'child_process';
import { join, resolve } from 'path';
import { validator } from '@liskhq/lisk-validator';

import { ApplicationConfig, applicationConfigSchema } from 'lisk-framework';
import { ConfigV3, Logger } from '../types';
import { NETWORK_CONSTANT } from '../constants';

const debug = debugInit('lisk:migrator');

const LOG_LEVEL_PRIORITY = Object.freeze({
	TRACE: 0,
	DEBUG: 1,
	INFO: 2,
	WARN: 3,
	ERROR: 4,
	FATAL: 5,
}) as Record<string, unknown>;

export const isBinaryBuild = (corePath: string): boolean => existsSync(join(corePath, '.build'));

export const getLogLevel = (logger: Logger): string => {
	const filteredLogLevelOptions = Object.keys(LOG_LEVEL_PRIORITY).reduce(
		(key: Record<string, unknown>, value: string) => {
			if (Object.values(logger).includes(value.toLowerCase())) {
				key[value] = LOG_LEVEL_PRIORITY[value];
			}
			return key;
		},
		{},
	);

	const logLevel = (Object.keys(filteredLogLevelOptions).find(
		key =>
			filteredLogLevelOptions[key] ===
			Math.min(...(Object.values(filteredLogLevelOptions) as Array<number>)),
	) as unknown) as string;

	return logLevel?.toLowerCase();
};

export const getConfig = async (corePath: string, customConfigPath?: string): Promise<ConfigV3> => {
	const command = [];

	const [network] = readdirSync(`${corePath}/config`);

	let compiledConfigPath = `${corePath}/config/${network}/config.json`;

	command.push(`cd ${corePath}`);

	if (isBinaryBuild(corePath)) {
		command.push('&& source env.sh');
	}

	if (customConfigPath) {
		command.push(`--config ${customConfigPath}`);
		compiledConfigPath = customConfigPath;
	}

	const fullCommand = command.join(' ');

	debug(`Core path: ${corePath}`);
	debug(`Cmd: ${fullCommand}`);

	cli.action.start('Compiling Lisk Core configuration');
	// Executing command to compile the configuration
	// 	to use the "source" command on Linux we have to explicity set shell to bash
	execSync(fullCommand, { shell: '/bin/bash' });
	cli.action.stop();

	cli.action.start('Loading Lisk Core configuration');
	const config = await import(compiledConfigPath);
	cli.action.stop();

	return config;
};

export const resolveConfigPathByNetworkID = async (networkIdentifier: string): Promise<string> => {
	const network = NETWORK_CONSTANT[networkIdentifier].name;
	const configFilePath = resolve(process.cwd(), `config/${network}/config.json`);
	return configFilePath;
};

export const createBackup = async (config: ConfigV3): Promise<void> => {
	const backupPath = `${process.cwd()}/backup`;
	mkdirSync(backupPath, { recursive: true });
	writeFileSync(resolve(`${backupPath}/config.json`), JSON.stringify(config, null, '\t'));
};

// TODO: Set up a default config file. Log properties and map migrated config values
export const migrateUserConfig = async (
	configv3: ConfigV3,
	configV4: ApplicationConfig,
): Promise<ApplicationConfig> => {
	if (configv3.rootPath) {
		cli.action.start(`Migrating config property 'dataPath' to ${configv3.rootPath}.`);
		configV4.system.dataPath = configv3.rootPath;
		cli.action.stop();
	}

	if (configv3.logger) {
		const logLevel = getLogLevel(configv3.logger);
		cli.action.start(`Migrating config property 'logLevel' to ${logLevel}.`);
		configV4.system.logLevel = logLevel;
		cli.action.stop();
	}

	if (configv3.transactionPool) {
		cli.action.start("Migrating config property 'transactionPool'.");
		configV4.transactionPool = configv3.transactionPool;
		cli.action.stop();
	}

	if (configv3.rpc) {
		cli.action.start(`Migrating config property 'rpc' mode to ${configv3.rpc.mode}.`);
		configV4.rpc.modes = [configv3.rpc.mode];
		cli.action.stop();
	}

	if (configv3.network) {
		if (configv3.network.port) {
			cli.action.start(`Migrating config property 'network' port to ${configv3.network.port}.`);
			configV4.network.port = configv3.network.port;
			cli.action.stop();
		}

		if (configv3.network.hostIp) {
			cli.action.start(`Migrating config property 'network' host to ${configv3.network.hostIp}.`);
			configV4.network.host = configv3.network.hostIp;
			cli.action.stop();
		}

		if (configv3.network.maxOutboundConnections) {
			cli.action.start(
				`Migrating config property 'network' maxOutboundConnections to ${configv3.network.maxOutboundConnections}.`,
			);
			configV4.network.maxOutboundConnections = configv3.network.maxOutboundConnections;
			cli.action.stop();
		}

		if (configv3.network.maxInboundConnections) {
			cli.action.start(
				`Migrating config property 'network' maxInboundConnections to ${configv3.network.maxInboundConnections}.`,
			);
			configV4.network.maxInboundConnections = configv3.network.maxInboundConnections;
			cli.action.stop();
		}

		if (configv3.network.wsMaxPayload) {
			cli.action.start(
				`Migrating config property 'network' wsMaxPayload to ${configv3.network.wsMaxPayload}.`,
			);
			configV4.network.wsMaxPayload = configv3.network.wsMaxPayload;
			cli.action.stop();
		}

		if (configv3.network.advertiseAddress) {
			cli.action.start(
				`Migrating config property 'network' advertiseAddress to ${configv3.network.advertiseAddress}.`,
			);
			configV4.network.advertiseAddress = configv3.network.advertiseAddress;
			cli.action.stop();
		}
	}

	return configV4;
};

export const validateConfig = async (config: ApplicationConfig): Promise<boolean> => {
	try {
		(await validator.validate(applicationConfigSchema, config)) as unknown;
		return true;
	} catch (_) {
		return false;
	}
};

export const writeConfig = async (config: ApplicationConfig, outputPath: string): Promise<void> => {
	if (existsSync(outputPath)) {
		unlinkSync(outputPath);
	}

	mkdirSync(outputPath, { recursive: true });

	writeFileSync(resolve(outputPath, 'config.json'), JSON.stringify(config));
};
