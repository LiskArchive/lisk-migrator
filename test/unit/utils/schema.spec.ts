import { defaultAccountSchema } from '../../../src/utils/schema';

describe('utils/schema', () => {
	describe('defaultAccountSchema', () => {
		it('should match the snapshot', () => {
			expect(defaultAccountSchema).toMatchSnapshot();
		});
	});
});
