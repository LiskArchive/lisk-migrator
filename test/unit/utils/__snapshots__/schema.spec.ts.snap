// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/schema defaultAccountSchema should match the snapshot 1`] = `
Object {
  "dpos": Object {
    "default": Object {
      "delegate": Object {
        "consecutiveMissedBlocks": 0,
        "isBanned": false,
        "lastForgedHeight": 0,
        "pomHeights": Array [],
        "totalVotesReceived": 0n,
        "username": "",
      },
      "sentVotes": Array [],
      "unlocking": Array [],
    },
    "fieldNumber": 5,
    "properties": Object {
      "delegate": Object {
        "fieldNumber": 1,
        "properties": Object {
          "consecutiveMissedBlocks": Object {
            "dataType": "uint32",
            "fieldNumber": 3,
          },
          "isBanned": Object {
            "dataType": "boolean",
            "fieldNumber": 5,
          },
          "lastForgedHeight": Object {
            "dataType": "uint32",
            "fieldNumber": 4,
          },
          "pomHeights": Object {
            "fieldNumber": 2,
            "items": Object {
              "dataType": "uint32",
            },
            "type": "array",
          },
          "totalVotesReceived": Object {
            "dataType": "uint64",
            "fieldNumber": 6,
          },
          "username": Object {
            "dataType": "string",
            "fieldNumber": 1,
          },
        },
        "required": Array [
          "username",
          "pomHeights",
          "consecutiveMissedBlocks",
          "lastForgedHeight",
          "isBanned",
          "totalVotesReceived",
        ],
        "type": "object",
      },
      "sentVotes": Object {
        "fieldNumber": 2,
        "items": Object {
          "properties": Object {
            "amount": Object {
              "dataType": "uint64",
              "fieldNumber": 2,
            },
            "delegateAddress": Object {
              "dataType": "bytes",
              "fieldNumber": 1,
            },
          },
          "required": Array [
            "delegateAddress",
            "amount",
          ],
          "type": "object",
        },
        "type": "array",
      },
      "unlocking": Object {
        "fieldNumber": 3,
        "items": Object {
          "properties": Object {
            "amount": Object {
              "dataType": "uint64",
              "fieldNumber": 2,
            },
            "delegateAddress": Object {
              "dataType": "bytes",
              "fieldNumber": 1,
            },
            "unvoteHeight": Object {
              "dataType": "uint32",
              "fieldNumber": 3,
            },
          },
          "required": Array [
            "delegateAddress",
            "amount",
            "unvoteHeight",
          ],
          "type": "object",
        },
        "type": "array",
      },
    },
    "type": "object",
  },
  "keys": Object {
    "default": Object {
      "mandatoryKeys": Array [],
      "numberOfSignatures": 0,
      "optionalKeys": Array [],
    },
    "fieldNumber": 4,
    "properties": Object {
      "mandatoryKeys": Object {
        "fieldNumber": 2,
        "items": Object {
          "dataType": "bytes",
        },
        "type": "array",
      },
      "numberOfSignatures": Object {
        "dataType": "uint32",
        "fieldNumber": 1,
      },
      "optionalKeys": Object {
        "fieldNumber": 3,
        "items": Object {
          "dataType": "bytes",
        },
        "type": "array",
      },
    },
    "type": "object",
  },
  "sequence": Object {
    "default": Object {
      "nonce": 0n,
    },
    "fieldNumber": 3,
    "properties": Object {
      "nonce": Object {
        "dataType": "uint64",
        "fieldNumber": 1,
      },
    },
    "type": "object",
  },
  "token": Object {
    "default": Object {
      "balance": 0n,
    },
    "fieldNumber": 2,
    "properties": Object {
      "balance": Object {
        "dataType": "uint64",
        "fieldNumber": 1,
      },
    },
    "type": "object",
  },
}
`;
